Expose extra entities for Nimly Locks

uvNikita opened the pull request:
## Proposed change

- Last action source
- Last action
- Last action user
- Last pin code
- Auto relock
- Sound volume

![image](https://github.com/user-attachments/assets/fad04c38-bd79-4118-b58f-f7f7c9458c1f)


## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

Fixes #3095
Fixes #2354

Related to #3465, #3580, #3457


## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

uvNikita committed:
Publish extra entities for Nimly Locks

- Last action source
- Last action
- Last action user
- Last pin code
- Auto relock
- Sound volume

TheJulianJES commented on zhaquirks/nimly/lock.py:
```diff
@@ -25,11 +32,121 @@
+def last_action_source_converter(value: int) -> str:
+    """Extract last action source value."""
+    value_hex = hex(value)[2:].zfill(8)
+    sources = {
+        "00": "Zigbee",
+        "02": "Keypad",
+        "03": "Fingerprint",
+        "04": "RFID",
+        "0a": "Self",
```
I think we should consider using `snake_case` for the enum sensor values, so just "zigbee", "rfid", and so on. We should be able to translate the entries by adding them to `strings.json` in HA (manually) then.

uvNikita replied:
Done

TheJulianJES commented on zhaquirks/nimly/lock.py:
```diff
@@ -25,11 +32,121 @@
+        "04": "RFID",
+        "0a": "Self",
+    }
+    return sources.get(value_hex[0:2], "Unknown")
```
I think returning `None` should mark the sensor as "Unavailable", where it'll go grey.
But we can also keep "unknown" if you prefer that.

uvNikita replied:
Done

TheJulianJES reviewed (commented):
Also, can you write a very small test that just calls the uncovered functions used as `attribute_converter`s for the quirks v2 entities? Just make sure the output is expected.

uvNikita commented:
> Also, can you write a very small test that just calls the uncovered functions used as `attribute_converter`s for the quirks v2 entities? Just make sure the output is expected.

Good point, added tests.

Also fixed other comments and changed `last_pin_code` to initially disabled since it might be a bit sensitive information.

TheJulianJES commented on zhaquirks/nimly/lock.py:
```diff
@@ -25,11 +32,121 @@
+        endpoint_id=11,
+        cluster_id=NimlyDoorLock.cluster_id,
+        attribute_name=NimlyDoorLock.AttributeDefs.nimly_last_lock_unlock_source.name,
+        translation_key="last_action_source",
+        fallback_name="Last action source",
```
Can we move `translation_key` and then `fallback_name` to the bottom (so as the last argument per method)?
That matches what we do for all other v2 quirks.

uvNikita replied:
Done

TheJulianJES commented on zhaquirks/nimly/lock.py:
```diff
@@ -25,11 +32,121 @@
+    return user_id
+
+
+def last_pin_code_converter(value: bytes) -> str:
+    """Extract last used pin code."""
+    return "".join(hex(b)[2:] for b in value)
```
Can you also add one small test for this method?

puddly replied:
I believe this implementation may have a bug:

```python
In [5]: value = b'\x04\x00\x05XI=\x1bg\x0f3'

In [6]: value.hex()
Out[6]: '04000558493d1b670f33'

In [7]: "".join(hex(b)[2:] for b in value)
Out[7]: '40558493d1b67f33'
```

You may be able to use `lambda value: value.hex()` as the attribute converter.

uvNikita replied:
@puddly so, if I understand correctly, my implementation skips zeros if they are in the first part of the byte? I think you are right. Changed to just `.hex()` call as you suggested. 

@TheJulianJES then there is no need for test I guess?

TheJulianJES replied:
Since it's inline then, yeah, not really. It's still not covered technically, but it's very simple now, so it's fine IMO.

TheJulianJES reviewed (approved):
Thanks! I think this looks good now.

We'll need to add the enum sensor translations to Home Assistant directly then. I might be able to do it in the bump PR, but we'll see. It's not completely required to do together, since they'll just show up without the translations (in lower-case) otherwise.

Another note: You don't need to squash and force-push with every change/commit. We squash all PR commits before merging anyway.

And one last thing, you can also use `.clone()` for base quirks if you have one base quirk (e.g. with quirks v2 entities) and then want to share them across multiple v2 quirks for different Zigbee devices/models.
See this example: https://github.com/zigpy/zha-device-handlers/blob/e3124b5504c6f0ae0fc1a20969fa46a5fee0c94d/zhaquirks/tuya/tuya_co.py#L38-L84

It's a simple implementation, but I think it should even work here if you wish to use that in a future PR.

uvNikita commented:
> And one last thing, you can also use `.clone()` for base quirks if you have one base quirk (e.g. with quirks v2 entities) and then want to share them across multiple v2 quirks for different Zigbee devices/models.

That's nice to know, I was wondering if it was available. Will use it in the future :+1: 

Thanks for the review!

TheJulianJES commented:
@uvNikita Quick question:
Working on the `strings.json` part right now. Do you want "Lock" and "Unlock" for "Last action" or are "Locked" and "Unlocked" also fine/better?

Example:
```json
      "last_action_source": {
        "name": "Last action source",
        "state": {
          "zigbee": "Zigbee",
          "keypad": "Keypad",
          "fingerprint": "Fingerprint",
          "rfid": "RFID",
          "self": "Self"
        }
      },
      "last_action": {
        "name": "Last action",
        "state": {
          "lock": "[%key:common::state::locked%]",
          "unlock": "[%key:common::state::unlocked%]"
        }
      },
```

(The common part comes from here https://github.com/home-assistant/core/blob/7b8ebb08034044ee3adc282213846afc8b8097d6/homeassistant/strings.json#L137 and is technically for the current state, but these common parts prevents duplicating translations. We can also just use "Lock" or "Unlock" if you prefer that.)

uvNikita commented:
I think locked/unlocked works too if it helps to reduce translations 👍

TheJulianJES commented:
Ok, I'll use those in the bump PR. We can always adjust later if that's not ideal.
But I guess "Unlocked" and "Locked" also kinda make sense for the last action because it already happened..? 😄

uvNikita commented:
Exactly 😅

f03el commented:
Great! Nice to see better support for Nimly locks. Just one thing: will managing pincodes/users be possible with this change? That's how I interpret #3095.
